### PR TITLE
Implement frontend login logic

### PR DESCRIPTION
**Pull request by aikenwx**
- Set up frontend routing and basic log in function
- Fixes bug for database setup script
- Fixes unknown module bug for frontend
- #4 #23

### CODE REVIEW

File: frontend/src/App.tsx [hunk 1]
```diff
@@ -1,8 +1,37 @@
...
+  };
+
+  const signIn = (auth: IAuth) => {
+    setAuthState(auth);
+  };
+
+  const value = { authState, redirectToSignIn, signIn, signout };
+
+  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
+}
```
Comment (chownces) [inline frontend/src/App.tsx]: Let's abstract these away into another file(s), together with the interfaces. Maybe under frontend/providers or frontend/contextProviders

File: frontend/src/pages/AuthPage.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { useEffect } from 'react';
```
Comment (chownces) [inline frontend/src/pages/AuthPage.tsx]: Let's rename this file to AuthRedirect

File: frontend/src/routes/RequireAuth.tsx [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { Navigate, Outlet, useLocation } from 'react-router-dom';
+import { useAuth } from '../utils/hooks';
+
+export default function RequireAuth() {
```
Comment (chownces) [inline frontend/src/routes/RequireAuth.tsx]: Let's rename this to `ProtectedRoutes`

File: frontend/src/routes/PublicOnly.tsx [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import { Navigate, useLocation, Outlet } from 'react-router-dom';
+import { useAuth } from '../utils/hooks';
+
+export default function PublicOnly() {
```
Comment (chownces) [inline frontend/src/routes/PublicOnly.tsx]: Let's rename this to `PublicOnlyRoutes`

File: frontend/src/routes/index.tsx [hunk 5]
```diff
@@ -1,15 +1,45 @@
...
-export default createBrowserRouter([
+export const router = createBrowserRouter([
   {
-    path: '/login',
-    element: <Login />,
+    element: (
+      <>
+        <NavBar />
+        <Outlet />
+      </>
```
Comment (chownces) [inline frontend/src/routes/index.tsx]: Let's abstract this into App.tsx or something

File: frontend/src/App.tsx [hunk 6]
```diff
@@ -1,8 +1,13 @@
-import Router from './routes';
+// import Login from './pages/Login';
```
Comment (chownces) [inline frontend/src/App.tsx:1]: Unused comment

File: frontend/src/contextProviders/AuthContext.tsx [hunk 7]
```diff
@@ -0,0 +1,27 @@
+// import Login from './pages/Login';
```
Comment (chownces) [inline frontend/src/contextProviders/AuthContext.tsx:1]: Unused comment

File: frontend/src/contextProviders/AuthContext.tsx [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import { IAuth, IAuthContext } from '../interfaces';
+import { useLocalStorageState } from '../utils/hooks';
+
+export const AuthContext = React.createContext<IAuthContext>(null!);
+
+export function AuthProvider({ children }: { children: React.ReactNode }) {
+  const [authState, setAuthState] = useLocalStorageState<IAuth | null>('authState', null);
+
+  const redirectToSignIn = () => {
+    window.location.replace(`${import.meta.env.VITE_BACKEND_API_HOST}/v1/auth/google`);
```
Comment (chownces) [inline frontend/src/contextProviders/AuthContext.tsx:13]: We can consider using a utils/Constants.ts file to consolidate these env var imports in the future

Comment (chownces) [inline frontend/src/contextProviders/AuthContext.tsx:13]: We can also consider having a backApi.ts file that has an object containing all the backend routes

```
const AUTH = '/auth';
const QUESTION = '/question';
const backendApi = {
  auth: {
    root: AUTH, // doesn't exist. just for example
    google: `${AUTH}/google`
  },
  question: {
    ...
  }
}

File: frontend/src/pages/AuthRedirect.tsx [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  useEffect(() => {
+    const params = new URLSearchParams(window.location.search);
+
+    const accessToken = params.get('accessToken');
+    const refreshToken = params.get('refreshToken');
+
+    auth.signIn({
+      accessToken: accessToken!,
+      refreshToken: refreshToken!,
+    });
```
Comment (chownces) [inline frontend/src/pages/AuthRedirect.tsx:19]: Let's add some error handling here incase either token is missing

**Review by chownces — APPROVED** [review]
Some more minor comments. Feel free to address them here or in a separate PR

File: frontend/src/pages/Login.tsx [hunk 10]
```diff
@@ -13,9 +16,7 @@ export default function Login() {
         <Typography variant="h1" gutterBottom>
           PeerPrep
         </Typography>
-        <TextField id="outlined-basic" label="Username" variant="outlined" />
-        <TextField id="outlined-basic" label="Password" variant="outlined" />
-        <Button href={`/dashboard`} variant="contained">
+        <Button onClick={auth.redirectToSignIn} variant="contained">
```
Comment (kwokieee) [inline frontend/src/pages/Login.tsx:19]: Can we remove this now that the NavBar contains the login/logout button?

File: frontend/vite.config.ts [hunk 11]
```diff
@@ -4,6 +4,9 @@ import checker from 'vite-plugin-checker'
 
 // https://vitejs.dev/config/
 export default defineConfig({
+  server: {
```
Comment (kwokieee) [inline frontend/vite.config.ts:7]: In addition, can we add the following lines to ensure that the build preview will also be run on the same port?

```
preview: {
  3000
}
```

File: frontend/src/App.tsx (hunk 6 above)
Comment (aikenwx) [reply frontend/src/App.tsx:1]: Will be addressed in separate PR

File: frontend/src/contextProviders/AuthContext.tsx (hunk 7 above)
Comment (aikenwx) [reply frontend/src/contextProviders/AuthContext.tsx:1]: Will be addressed in separate PR

File: frontend/src/contextProviders/AuthContext.tsx (hunk 8 above)
Comment (aikenwx) [reply frontend/src/contextProviders/AuthContext.tsx:13]: Will be addressed in separate PR

Comment (aikenwx) [reply frontend/src/contextProviders/AuthContext.tsx:13]: Will be addressed in separate PR

File: frontend/src/pages/AuthRedirect.tsx (hunk 9 above)
Comment (aikenwx) [reply frontend/src/pages/AuthRedirect.tsx:19]: Will be addressed in separate PR

File: frontend/src/pages/Login.tsx (hunk 10 above)
Comment (aikenwx) [reply frontend/src/pages/Login.tsx:19]: Will be addressed in separate PR

File: frontend/vite.config.ts (hunk 11 above)
Comment (aikenwx) [reply frontend/vite.config.ts:7]: Will be addressed in separate PR

**Pull request merged**